app/gui2/src/workers/visualizationCompiler.ts

```diff
@@ -0,0 +1,298 @@
+ * URL (`/visualizations/TheScript.vue`). Relative URLs (`./TheScript.vue`) are NOT valid.
+ *
+ * Note that compiling files other than Vue files (TypeScript, SVG etc.) are currently NOT
+ * supported.
+ * ```ts
+ * interface CompileRequest {
+ *   type: 'compile'
+ *   id: number
+ *   path: string
+ * }
+ * ```
+ *
+ * ### Register Builtin Modules
+ * This is a request to mark modules as built-in, indicating that the compiler should re-write the
+ * imports into object destructures.
+ * ```ts
+ * interface RegisterBuiltinModulesRequest {
+ *   type: 'register-builtin-modules'
+ *   modules: string[]
+ * }
+ * ```
+ *
+ * ## Worker to Main Thread
+ *
+ * ### Compilation Result
+ * Sent in response to a `CompileRequest`, with an `id` matching the `id` of the original request.
+ * Contains only the `path` of the resulting file (which should have also been sent in the
+ * `CompileRequest`). The content itself will have been sent earlier as an `AddImportNotification`.
+ * ```ts
+ * interface CompilationResultResponse {
+ *   type: 'compilation-result'
+ *   id: number
+ *   path: string
+ * }
+ *
+ * ### Add Style
+ * Sent after compiling `<style>` and `<style scoped>` sections. These should be attached to the
+ * DOM - placement does not matter.
+ * ```ts
+ * interface AddStyleNotification {
+ *   type: 'style'
+ *   code: string
+ * }
+ * ```
+ *
+ * ### Add Raw Import
+ * Currently unused.
+ *
+ * Sent after compiling an import which does not result in a URL.
+ *
+ * Should be added to the cache using `cache[path] = value`.
+ * ```ts
+ * interface AddRawImportNotification {
+ *   type: 'raw-import'
+ *   path: string
+ *   value: unknown
+ * }
+ * ```
+ *
+ * ### Add URL Import
+ * Sent after compiling an import which results in a URL as its default export.
+ * This is usually the case for assets.
+ *
+ * Should be added to the cache using `cache[path] = { default: URL.createObjectURL(new Blob([value], { type: mimeType })) }`.
+ * ```ts
+ * interface AddURLImportNotification {
+ *   type: 'url-import'
+ *   path: string
+ *   mimeType: string
+ *   value: BlobPart
+ * }
+ * ```
+ *
+ * ### Add Import
+ * Sent after compiling a JavaScript import.
+ *
+ * Should be added to the cache using `cache[path] = import(URL.createObjectURL(new Blob([code], { type: 'text/javascript' })))`.
+ * ```ts
+ * interface AddImportNotification {
+ *   type: 'import'
+ *   path: string
+ *   code: string
+ * }
+ * ```
+ */
+
+import { parse as babelParse } from '@babel/parser'
+import MagicString from 'magic-string'
+import { transform } from 'sucrase'
+import { compileScript, compileStyle, parse } from 'vue/compiler-sfc'
+
+let builtinModules = new Set<string>()
+const ids = new Set<string>()
+function generateId() {
+  for (;;) {
+    const id = Math.floor(Math.random() * 0xffffffff)
+      .toString(16)
+      .padStart(8, '0')
+    if (!ids.has(id)) {
+      return id
+    }
+  }
+}
```

Comment:
This can be replaced with `random.uuidv4` from `lib0`.

Reply:
results of discord discussion: we'll use hash-sum because that's what `compiler-sfc` seems to use internally